# 推荐规则  
-v1.2 李昱辰  
1.添加用户影响因素  
2.添加用户联系系数  
3.修正模型为可控模型
  

### 一. 用户标签规则  
##### 规则一：每个用户的好友(或粉丝)中与该用户具有相同兴趣的人占多数  
##### 规则二：与用户越相似的好友或粉丝，其标签越可能是用户的标签  
##### 规则三：用户与某好友互动频率越高，用户与该好友的兴趣相似度越大（包括评论、动态、聊天）  

因为APP注册过程并没有强制性的添加兴趣爱好，所以用户一些相对标签无法获得，只能获得用户地址，学校，专业，毕业时间等。  
这个时候需要给用户一些默认的标签，按照用户所在学校的人群大类兴趣，从学校的用户的已有数据，进行推导，把大类标签作为用户的默认标签。

1.学校标签  
通过学校的共同性进行推荐，清华的学生更多地推荐清华的学生  

2.专业标签  
通过专业的共同性进行推荐，但是因为有很多交叉学科（比如UI设计就是设计和心理学、计算机的交叉学科，这样可以给用户打上不同的用户标签，推荐多个设计、心理学、计算机的个集或者交集的用户）、相同大类用户（比如文科类用户推荐文科用户、理科类用户推荐理科用户、工科推荐工科或者理科用户） 

3.学院标签
通过学院的共同性进行推荐，经济学院的学生更容易和经济学院的学生形成交流

3.年限推荐  
大学四年时间内，所接触到的各个年级学生，更容易成为密集交流者  
  
4.兴趣推荐  
1)用户设定兴趣的推荐  
  通过用户自己设定的兴趣进行推荐，但是因为用户的“设定行为”具有不确定性，所以对于用户的准确的兴趣有一定的标准差。  
2)用户行为兴趣的推荐  
  ①用户好友的兴趣标签：取出现最多的几个标签作为用户隐性标签。  
  ②用户参与活动的兴趣标签：因为活动没有设定标签，可以通过参与用户中出现最多的几个标签，作为活动的标签，这样也可返回为用户兴趣标签。

### 二.用户价值
用户价值暂时定为用户使用频率（单日内打开次数）、用户使用时长（单日内所有时长）、用户影响指数（动态被参与次数，如转发、评论、点赞等）  

#### 1.使用频率与用户使用时长  
用以衡定用户活跃度，如果长时间没有使用，那么用户对于平台的价值会下降。  

#### 2.用户影响指数  
##### 规则一：影响力越高的用户回复的动态的影响力越高，从而使该动态主人的影响力变高。  
  
##### 规则二：影响力越高的用户转发的动态的影响力越高，从而使该动态原创作者的影响力变高。  
  
##### 规则三：影响力越高的用户倾向于在其动态中@影响力高或者的用户。  
  
PageRank算法：  
1.赋予所有用户相同的权重。  
2.将每个用户的影响力权重按照其关注的人数等量分配  
3.对每个用户来说，其影响力等于其粉丝分配给他的权重之和  
4.第2步和第3步迭代，直到权重不再发生大的变化为止  

运用pagerank算法，可以计算出不同的用户，不同的影响力权重。
但是这个权重有很大的不稳定性，如果有人因为一些原因，关注的人和好友突涨，那么整体的影响力会上升特别快。
用户的影响力除了他的动态关系之外，还与他的个人属性有很大的关系，比如动态的质量。动态的质量可以采用其被转发的数目、被回复的数目、被点赞的数目来得到。

### 三.用户相似度
运用余弦相似度来衡定用户的相似性，加入权重因素，让衡定结果更准确。  

##### 1.余弦相似度
余弦夹角公式： 
$$ \cos\theta = \frac{\alpha*\beta}{|\alpha|·|\beta|} $$  

公式变形后：
$$ \cos\theta = \frac{\alpha*\beta}{\sqrt{\alpha^2}·\sqrt{\beta^2}} $$

运用：   
  
 用户  |  学校  |  专业  | 兴趣大类 | 兴趣小类  
user1 | 锦城大学| 经济学 |   体育   |  足球  
user2 | 锦城大学| 保险学 |   体育   |  足球  

转换为用户矩阵：  
 用户  |  锦城大学  |  经济学  |  保险学  |  体育  |  足球  
user1 |      1    |    1    |    0    |   1   |    1  
user2 |      1    |    0    |    1    |   1   |    1  

则两个用户的相似度为：
$$\because \cos\theta = \frac{1\times1+1\times0+0\times1+1\times1+1\times1}{\sqrt{1^2+1^2+0^2+1^2+1^2}\times\sqrt{1^2+0^2+1^2+1^2+1^2}} $$  

$$\therefore \cos\theta = \frac{3}{4} = 0.75$$  

权重算法：对于大类加上权重，比如锦城大学、比如体育  
 用户  |  锦城大学  |  经济学  |  保险学  |  体育  |  足球  
user1 |      2    |    1    |    0    |   2   |    1  
user2 |      2    |    0    |    1    |   2   |    1  

$$\because \cos\theta = \frac{2\times2+1\times0+0\times1+2\times2+1\times1}{\sqrt{2^2+1^2+0^2+2^2+1^2}\times\sqrt{2^2+0^2+1^2+2^2+1^2}} $$  

$$\therefore \cos\theta = \frac{9}{10} = 0.9$$  
  
加上权重过后，相同的会表现出更好的相同性，但是不同的，就会有更大的差异性。  
##### 2.余弦相似度变形  
对于公式再理解，可以看出其实就是相同的特征在所有的特征中的占比。  

$$\because \cos\theta = \frac{相同的特征}{所有特征} $$  
$$\therefore \cos\theta = \frac{user1_s\bigcap user2_s}{user1_s\bigcup user2_s}$$

### 四.用户联系系数  
##### 1.用户互动系数  
用户互动系数为不同用户之间的交流系数，包括用户的所有互动（交流、评论、点赞、转发、参与）  

$$ 
设 交流为X_1、评论为X_2、转发为X_3、点赞为X_4、参与X_5,  
\therefore \lambda_1>\lambda_2>\lambda_3>\lambda_4>\lambda_5
$$  
$$
用户互动系数 = \log\sum^n(\lambda_1 X_1 + \lambda_2 X_2+\lambda_3 X_3+\lambda_4 X_4+\lambda_5 X_5)
$$  

##### 2.用户联系系数 （用户互动相似系数）
用户A有两个互动系数较高用户B与用户C，用户B和用户A的互动系数等于30，用户C与用户A的互动系数等于31，则用户B到用户A的互动相似系数为30/31=0.96   
$$  
设 A\to B = 30,C\to B =31
$$  
$$
用户联系系数 = \frac{min(A\to B,C\to B)}{max(A\to B,C\to B)}
$$   

这样推荐出来的用户，更容易三人为团，形成一个社交圈  

### 五.推荐系数  
推荐系数的构建包含了用户价值、用户联系系数、用户相似度三个参数的结合，根据不同的运营场景，可以调整参数的权重用以适应。  
  
## $$推荐系数 = \log(\lambda_1 * 用户联系系数 + \lambda_2 * 用户价值 + \lambda_3 * 用户相似度)$$

# Q&A
#### Q：1.用户标签如何标记和构建？  
  
A：初始的时候会根据用户的填写，如学校、专业、学年来构建特征、用户也可以在兴趣爱好里填写自己的兴趣爱好，但考虑到兴趣爱好的变化性、用户资料的主观性以及填写信息的时间敏感性，则按照用户行为（如用户点赞、评论、转发、参与动态、参与活动、交友圈的隐性标签作为用户较为客观的）来构建。  
  
#### Q：2.用户价值为什么要引入用户影响指数？  
  
A：用户影响指数是要在APP中推荐发文质量较高，关注人数较多的用户，这样的用户可以带动整个平台的质量和活跃性，也可以激活普通用户成为“大V”用户。  
#### Q：3.用户联系系数的意义在哪？  
  
A：比如A有B和C两个密切联系好友，那么B和C是有互相认识的意义的，这样有助于生成更加密切用户社交网络，而不是散开的形式。  
  
#### Q：4. 推荐系数的三个权重参数如何进行选择？   
  
A：按照运营策略进行选择：  
   (1)如果是推进高质量用户、以及主打质量内容，则可以主要加大用户价值的权重。  
   (2)如果在前期，用户好友较少，单用户的交流互动较少，则可以主要加大用户相似度的权重。  
   (3)如果平台要形成用户交流圈子，以及刺激一些线下活动的活跃（普遍人群号召比个人意愿更会参加活动板块），则可以主要加大用户联系系数的权重。  

# 示例

python 3.6.2 -jupyter notebook

In [1]:
#导入所需要的包
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

#读取数据，并对数据做初始处理
user_data = pd.read_csv('../data/user_data.csv')
user_data = user_data.drop('Unnamed: 0',axis=1)

#### 创建模拟用户

In [2]:
userA = pd.Series([
    101,
    '四川',
    '成都市',
    '西华大学',
    '经济学院',
    '文学',
    '经济学',
    '影视',
    '美丽人生',
    '2011',
    '技能',
    '编程',
    '名人',
    '搞笑漫画日和',
    '宋冬野',
    '搞笑漫画日和',
    '妖精的尾巴',
    '七龙珠',
    '搞笑漫画日和'
])

构建虚拟用户互动系数和用户价值  
- 用户系数 1-9
- 用户价值 1-9

In [3]:
user_data['similarity'] = ''
score_list = []
userA_set = set(userA)
userB_set = set(userA)

for i in range(0,len(user_data)):
    userB_set = set(user_data.loc[i])
    score = len(userA_set&userB_set)/(np.sqrt(len(userA_set))*np.sqrt(len(userB_set)))
    score_list.append(score)
    
user_data['similarity'] = score_list

user_data['value'] = ''
user_data['value'] = user_data['value'].apply(lambda x : np.random.choice(range(1,10)))

user_data['connect'] = ''
user_data['connect'] = user_data['connect'].apply(lambda x : np.random.choice(range(1,10)))

### $$ 推荐系数 = \log(\lambda_1 * 互动系数 + \lambda_2 * 用户价值 + \lambda_3 * 用户相似度) $$

In [5]:
user_data['score'] = ''
for i in range(0,len(user_data)):
    #核心算法，权重可修改进行推荐，现设置互动系数权重为0.2、用户价值权重位0.2、用户相似度权重为0.2
    user_data['score'][i] = np.log10(0.2*user_data['connect'][i]+0.2*user_data['value'][i]+0.2*60*user_data['similarity'][i])

In [6]:
user_data.sort_values('score',ascending=False)

id region_parent region_detail school_parent school_detail  \
154  10154            四川           成都市          西华大学          经济学院   
69   10069            四川           成都市        四川师范大学          经济学院   
232  10232            四川           成都市          西华大学          管理学院   
521  10521            四川           成都市        西南石油大学          经济学院   
213  10213            四川           内江市        西南财经大学          经济学院   
996  10996            四川           成都市        成都理工大学     土木与环境工程学院   
713  10713            四川           雅安市      成都信息工程大学          经济学院   
348  10348            四川           眉山市          西华大学          经济学院   
799  10799            四川           达州市        四川师范大学         计算机学院   
104  10104            四川           成都市        四川师范大学          艺术学院   
51   10051            四川           雅安市          西华大学          经济学院   
4    10004            四川           德阳市      成都信息工程大学          经济学院   
148  10148            四川           德阳市        电子科技大学          经济学院   
649  10649            四川          攀枝花市          四川大学          经济学院   
829  10829            四川           成都市          西华大学          经济学院   
992  10992            四川           自贡市      成都信息工程大学       文学与传媒学院   
156  10156            四川           广元市      成都信息工程大学          经济学院   
744  10744            四川          攀枝花市        西南财经大学          管理学院   
330  10330            四川           成都市        电子科技大学          经济学院   
558  10558            四川           雅安市          西华大学        财务会计学院   
985  10985            四川           成都市        成都理工大学          经济学院   
360  10360            四川           宜宾市        四川师范大学          经济学院   
704  10704            四川           乐山市        电子科技大学        财务会计学院   
963  10963            四川           自贡市        西南交通大学        财务会计学院   
737  10737            四川           乐山市        西南交通大学          经济学院   
451  10451            四川           达州市      四川大学锦城学院          经济学院   
636  10636            四川           遂宁市          西华大学     土木与环境工程学院   
216  10216            四川           德阳市        成都理工大学          管理学院   
313  10313            四川           绵阳市          西华大学          电气学院   
880  10880            四川           遂宁市        西南石油大学          管理学院   
..     ...           ...           ...           ...           ...   
581  10581            四川           南充市        西南石油大学          管理学院   
701  10701            四川           泸州市          西华大学     土木与环境工程学院   
946  10946            四川           广元市        西南财经大学         计算机学院   
719  10719            四川           宜宾市        成都理工大学          管理学院   
598  10598            四川           绵阳市          四川大学         计算机学院   
268  10268            重庆            重庆          四川大学          管理学院   
327  10327            重庆            重庆        西南石油大学        财务会计学院   
200  10200            重庆            重庆          四川大学        财务会计学院   
752  10752            四川           南充市      成都信息工程大学          管理学院   
679  10679            四川           宜宾市        四川师范大学          管理学院   
510  10510            四川           南充市        四川师范大学          管理学院   
58   10058            四川           南充市          四川大学         计算机学院   
583  10583            四川           广元市        西南石油大学          管理学院   
468  10468            重庆            重庆        西南石油大学     土木与环境工程学院   
542  10542            重庆            重庆      四川大学锦城学院          机械学院   
703  10703            四川           巴中市        成都理工大学          管理学院   
167  10167            四川           遂宁市        四川师范大学     土木与环境工程学院   
132  10132            四川           宜宾市        西南交通大学         计算机学院   
347  10347            重庆            重庆        西南石油大学          建筑学院   
367  10367            四川           自贡市          四川大学          管理学院   
657  10657            四川           乐山市        西南交通大学     土木与环境工程学院   
1    10001            四川           眉山市        西南交通大学     土木与环境工程学院   
41   10041            四川           资阳市        西南石油大学          管理学院   
630  10630            四川           自贡市      成都信息工程大学          管理学院   
690  10690            四川           内江市          四川大学          管理学院   
170  10170        

### 结果
可以看出id为10154的用户和10213、用户10036的推荐系数更高  
10154的用户在相似度上与模拟用户更加接近，达到了0.647059，所以加权相加后，推荐系数为1.01556  
10213的用户和10036的用户在用户价值和用户联系系数上更高，分别是9、9，推荐系数分别为0.894055、0.881167